File: src/epr_reader/bindings/frame_state.py
from enum import Enum

from dataclasses import dataclass

class FrameStateType(Enum):
    Unknown = 0

    SlideCommentWindow = 1
    """医生的下诊断窗口"""

    SlideToolsWindow = 2
    """切片工具选择窗口"""

    SlideRulerToolWindow = 3
    """尺子工具"""

    SlideMarkPenToolWindow = 4
    """标记笔工具"""

    CellNucleiDetectionTool = 5
    """细胞核探测工具"""

    Next = 0xFF,


@dataclass 
class FrameState:
    type: FrameStateType

File: src/EasyPathology.Record/FrameStates/FrameState.cs
using System.Diagnostics;
using EasyPathology.Record.Interfaces;

namespace EasyPathology.Record.FrameStates;

public static class FrameState {
	private static readonly Dictionary<FrameStateType, Func<IFrameState>> RegisteredFrameStates = new() {
		{ FrameStateType.SlideCommentWindow, static () => new SlideCommentWindowFrameState() },
		{ FrameStateType.SlideToolsWindow, static () => new SlideToolsWindowFrameState() },
		{ FrameStateType.SlideRulerToolWindow, static () => new SlideRulerToolWindowFrameState() },
		{ FrameStateType.SlideMarkPenToolWindow, static () => new SlideMarkPenToolWindowFrameState() },
	};

	public static FrameStateType ReadFrameStateType(this BinaryReader br) {
		try {
			return (FrameStateType)br.ReadByte();
		} catch (InvalidCastException) {
			return FrameStateType.Unknown;
		}
	}

	public static IFrameState ReadFrameState(this BinaryReader br, int version) {
		if (RegisteredFrameStates.TryGetValue(br.ReadFrameStateType(), out var frameStateInitializer)) {
			var frameState = frameStateInitializer();
			frameState.Read(br, version);
			return frameState;
		}

		throw new FrameStateTypeUnknownException();
	}

	public static void WriteFrameStateHeader(this IFrameState frameState, BinaryWriter bw, int version) {
		Debug.Assert(frameState != null);
		bw.Write((byte)frameState.Type);
		bw.Write(frameState.GetLength(version));
	}
}